minify-html/mod.ts

```diff
@@ -0,0 +1,109 @@
+import { init, minify as minifyHTML } from "./deps.ts";
+import { merge } from "lume/core/utils.ts";
+import { Exception } from "lume/core/errors.ts";
+
+import type { Page, Site } from "lume/core.ts";
+
+export interface Options {
+  /** The list of extensions this plugin applies to. */
+  extensions: string[];
+  /** If enabled, HTML files will be minified. */
+  minifyHTML: boolean;
+  /**
+   * Default options for minify-html library
+   * Opened an issue with a request of adding an export for the options type definitions so we can use `minifyOptions: Partial<CFG>`.
+   * https://github.com/wilsonzlin/minify-html/issues/97
+   * until then we define them manually.
+   */
+  minifyOptions: {
+    /** Do not minify DOCTYPEs. Minified DOCTYPEs may not be spec compliant. */
+    do_not_minify_doctype?: boolean;
+    /** Ensure all unquoted attribute values in the output do not contain any characters prohibited by the WHATWG specification. */
+    ensure_spec_compliant_unquoted_attribute_values?: boolean;
+    /** Do not omit closing tags when possible. */
+    keep_closing_tags?: boolean;
+    /** Do not omit `<html>` and `<head>` opening tags when they don't have attributes. */
+    keep_html_and_head_opening_tags?: boolean;
+    /** Keep spaces between attributes when possible to conform to HTML standards. */
+    keep_spaces_between_attributes?: boolean;
+    /** Keep all comments. */
+    keep_comments?: boolean;
+    /** If enabled, content in `<script>` tags with a JS or no [MIME type](https://mimesniff.spec.whatwg.org/#javascript-mime-type) will be minified using [minify-js](https://github.com/wilsonzlin/minify-js). */
+    minify_js?: boolean;
+    /** If enabled, CSS in `<style>` tags and `style` attributes will be minified. */
+    minify_css?: boolean;
+    /** Remove all bangs. */
+    remove_bangs?: boolean;
+    /** Remove all processing_instructions. */
+    remove_processing_instructions?: boolean;
+  };
+}
+
+// Default options
+export const defaults: Options = {
+  extensions: [".html"],
+  minifyHTML: true,
+  minifyOptions: {
+    do_not_minify_doctype: false,
+    ensure_spec_compliant_unquoted_attribute_values: false,
+    keep_closing_tags: false,
+    keep_html_and_head_opening_tags: false,
+    keep_spaces_between_attributes: false,
+    keep_comments: false,
+    minify_js: false,
+    minify_css: false,
+    remove_bangs: false,
+    remove_processing_instructions: false,
+  },
+};
+
+// Init minify-html
+await init();
+
+/** A plugin to minify HTML, CSS & JavaScript files in Lume */
+export default function (userOptions?: Partial<Options>) {
+  const options = merge(defaults, userOptions);
+
+  const { minifyOptions } = options;
+
+  if (!options.minifyHTML) {
+    options.extensions = options.extensions.filter((ext: string) =>
+      ext !== ".html"
+    );
+  }
+
+  if (minifyOptions.minify_js && !options.extensions.includes(".js")) {
+    options.extensions = options.extensions.concat([".js"]);
```

Comment:
This is not needed because it's configurable by editing the `extensions` option directly:

```js
site.use(minifyHtml({
    extensions: [".html", ".js"]
}));
```

Reply:
Starting point, I rename minifyOptions to options:

The `options.minify_js` default is set `false` (also for minify-html itself). To enable it for `.js` or `.css` you MUST pass `minify_js: true` to the minify-html options. If you just would pass the `.js` (or `.css`) extension to the plugin, without setting `minify_js: true`, minify-html will not minify `.js` or `.css`.

I thought my approach is a good idea, because the user just needs to set `minify_js: true` and the plugin will add `.js` to extensions automatically.

With your approach we need todo:

```js
if (options.extensions.includes(".js")) {
  options.options.minify_js = true;
}
```

Since when the user does not set `options.minify_js: true`, and just adds the extension, minify-html will not minify `.js`.

Hope you understand what I mean :blush:

What you think?

Reply:
See comment:

When `extensions` includes `.css` and `.js`, minify-html will minify CSS and JavaScript files, even when `minify_css` and `minify_js` defaults set to `false`, since lume passes them to process. To make this work with inline `<style>` tags in HTML files, following is necessary when `minify_css` is not set to `true`